src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -150,13 +153,64 @@ private async Task<Document> ApplyFixesAsync(Document document, ImmutableArray<A
+            var insertSpans = allTextChanges
+                .GroupBy(change => change.Span)
+                .Select(changes => new TextSpan(changes.Key.Start, changes.Sum(change => change.NewText.Length)));
+
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
             var newText = text.WithChanges(orderedTextInserts);
             var newDocument = newProject.GetDocument(document.Id).WithText(newText);
 
+            // When imports are added to a code file that has no previous imports extra
+            // newlines are generated between each import because the fix is expecting to
+            // separate the imports from the rest of the code file. We need to format the
+            // imports to remove these extra newlines.
+            return await CleanUpNewLinesAsync(newDocument, insertSpans, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<Document> CleanUpNewLinesAsync(Document document, IEnumerable<TextSpan> insertSpans, CancellationToken cancellationToken)
+        {
+            var languageFormatter = document.GetLanguageService<ISyntaxFormattingService>();
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            var newDocument = document;
+
+            // Since imports can be added at both the CompilationUnit and the Namespace level,
+            // format each span individually so that we can retain each newline that was intended
+            // separate the import section from the other content.
+            foreach (var insertSpan in insertSpans)
+            {
+                newDocument = await CleanUpNewLinesAsync(newDocument, insertSpan, languageFormatter, options, cancellationToken).ConfigureAwait(false);
+            }
+
             return newDocument;
         }
 
+        private async Task<Document> CleanUpNewLinesAsync(Document document, TextSpan insertSpan, ISyntaxFormattingService languageFormatter, OptionSet options, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var textChanges = languageFormatter.Format(root, new[] { insertSpan }, options, new[] { new CleanUpNewLinesFormatter(text) }, cancellationToken).GetTextChanges();
+
+            // If there are no changes then, do less work.
+            if (textChanges.Count == 0)
+            {
+                return document;
+            }
+
+            // If there are changes then, this was a case where there were no
+            // previous imports statements. We need to retain the final extra
+            // newline because that separates the imports section from the rest
+            // of the code.
+            textChanges.RemoveAt(textChanges.Count - 1);
```

Comment:
Can we add any assert here that matches your comment above regarding the removed textChange?

Reply:
Something like `Debug.Assert(textChanges.Last().Span.End == insertSpan.End);`?